crates/reward/src/lib.rs

```diff
@@ -0,0 +1,269 @@
+//! # Reward Module
+//! Based on the [Scalable Reward Distribution](https://solmaz.io/2019/02/24/scalable-reward-changing/) algorithm.
+
+#![deny(warnings)]
+#![cfg_attr(test, feature(proc_macro_hygiene))]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+use codec::{Decode, Encode, EncodeLike, FullCodec};
+use frame_support::dispatch::DispatchError;
+use sp_arithmetic::{
+    traits::{AtLeast32BitUnsigned, UniqueSaturatedInto},
+    FixedPointNumber,
+};
+use sp_runtime::traits::{CheckedAdd, CheckedDiv, CheckedMul, CheckedSub, MaybeSerializeDeserialize, Zero};
+use sp_std::{convert::TryInto, fmt::Debug, vec::Vec};
+
+pub(crate) type Inner<T, I = ()> = <<T as Config<I>>::SignedFixedPoint as FixedPointNumber>::Inner;
+
+pub(crate) type SignedFixedPoint<T, I = ()> = <T as Config<I>>::SignedFixedPoint;
+
+pub use pallet::*;
+
+pub type BackingVault = pallet::Instance1;
+pub type IssuingVault = pallet::Instance2;
+pub type BackingRelayer = pallet::Instance3;
+pub type IssuingRelayer = pallet::Instance4;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+    use frame_support::pallet_prelude::*;
+
+    /// ## Configuration
+    /// The pallet's configuration trait.
+    #[pallet::config]
+    pub trait Config<I: 'static = ()>: frame_system::Config {
+        /// The overarching event type.
+        type Event: From<Event<Self, I>> + IsType<<Self as frame_system::Config>::Event>;
+
+        /// The balance type for a currency.
+        type Balance: AtLeast32BitUnsigned + FullCodec + Copy + MaybeSerializeDeserialize + Debug + Default;
```

Comment:
Maybe worth adding the full currency module here so that reward claims are actually trade-able? Let me know if that is a quick thing or rather big change. If big -> not in this PR.

Reply:
The rewards will still be tradable, the `Currency` type (as used in the `currency` pallet) just provides the methods required by the pallet to burn, mint or transfer the asset. The `reward` pallet as no such need to do this internally - it just needs to know the type of the `Balances` which is most often `u128`.